internal/service/elbv2/hosted_zone_id_data_source.go

```diff
@@ -0,0 +1,117 @@
+	endpoints.UsEast2RegionID:      "Z3AADJGX6KTTL2",
+	endpoints.UsGovEast1RegionID:   "Z166TLBEWOO7G0",
+	endpoints.UsGovWest1RegionID:   "Z33AYJ8TM3BH4J",
+	endpoints.UsWest1RegionID:      "Z368ELLRRE2KJ0",
+	endpoints.UsWest2RegionID:      "Z1H1FL5HABSF5",
+}
+
+// See https://docs.aws.amazon.com/general/latest/gr/elb.html#elb_region
+var HostedZoneIdPerRegionNLBMap = map[string]string{
+	endpoints.AfSouth1RegionID:     "Z203XCE67M25HM",
+	endpoints.ApEast1RegionID:      "Z12Y7K3UBGUAD1",
+	endpoints.ApNortheast1RegionID: "Z31USIVHYNEOWT",
+	endpoints.ApNortheast2RegionID: "ZIBE1TIR4HY56",
+	endpoints.ApNortheast3RegionID: "Z1GWIQ4HH19I5X",
+	endpoints.ApSouth1RegionID:     "ZVDDRBQ08TROA",
+	endpoints.ApSoutheast1RegionID: "ZKVM4W9LS7TM",
+	endpoints.ApSoutheast2RegionID: "ZCT6FZBF4DROD",
+	endpoints.ApSoutheast3RegionID: "Z01971771FYVNCOVWJU1G",
+	endpoints.CaCentral1RegionID:   "Z2EPGBW3API2WT",
+	endpoints.CnNorth1RegionID:     "Z3QFB96KMJ7ED6",
+	endpoints.CnNorthwest1RegionID: "ZQEIKTCZ8352D",
+	endpoints.EuCentral1RegionID:   "Z3F0SRJ5LGBH90",
+	endpoints.EuNorth1RegionID:     "Z1UDT6IFJ4EJM",
+	endpoints.EuSouth1RegionID:     "Z23146JA1KNAFP",
+	endpoints.EuWest1RegionID:      "Z2IFOLAFXWLO4F",
+	endpoints.EuWest2RegionID:      "ZD4D7Y8KGAS4G",
+	endpoints.EuWest3RegionID:      "Z1CMS0P5QUZ6D5",
+	endpoints.MeSouth1RegionID:     "Z3QSRYVP46NYYV",
+	endpoints.SaEast1RegionID:      "ZTK26PT1VY4CU",
+	endpoints.UsEast1RegionID:      "Z26RNL4JYFTOTI",
+	endpoints.UsEast2RegionID:      "ZLMOA37VPKANP",
+	endpoints.UsGovEast1RegionID:   "Z1ZSMQQ6Q24QQ8",
+	endpoints.UsGovWest1RegionID:   "ZMG1MZ2THAWF1",
+	endpoints.UsWest1RegionID:      "Z24FKFUX50B4VW",
+	endpoints.UsWest2RegionID:      "Z18D5FSROUN65G",
+}
+
+func DataSourceHostedZoneID() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceHostedZoneIDRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"load_balancer_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      elbv2.LoadBalancerTypeEnumApplication,
+				ValidateFunc: validation.StringInSlice([]string{"application", "network"}, false), // should only accept application or network
+			},
+		},
+	}
+}
+
+func dataSourceHostedZoneIDRead(d *schema.ResourceData, meta interface{}) error {
+	region := meta.(*conns.AWSClient).Region
+	if v, ok := d.GetOk("region"); ok {
+		region = v.(string)
+	}
+
+	var lbType string
+	lbType = elbv2.LoadBalancerTypeEnumApplication
+	if v, ok := d.GetOk("load_balancer_type"); ok {
+		lbType = v.(string)
+	}
+
+	if lbType == elbv2.LoadBalancerTypeEnumApplication {
+		if zoneId, ok := HostedZoneIdPerRegionMap[region]; ok {
+			d.SetId(zoneId)
+			return nil
+		}
+	} else if lbType == elbv2.LoadBalancerTypeEnumNetwork {
+		if zoneId, ok := HostedZoneIdPerRegionNLBMap[region]; ok {
+			d.SetId(zoneId)
+			return nil
+		}
+	}
+
+	return fmt.Errorf("Unknown region (%q)", region)
```

Comment:
perhaps this can be shifted up in the function to check if the `region` passed by a user is inside the map and if not, throw this error. Then for this return, we can change it to a `return nil`. This means we can remove the `return nil` in the conditional statements above

Reply:
changed to `return nil` and added region validation as well in [a2e453e](https://github.com/hashicorp/terraform-provider-aws/pull/24749/commits/a2e453e420e2a9a17a8029ca4e45aca5b4b83911)